### ISSUE TITLE
Apply Woodbury matrix identity to GP

### ISSUE THREAD

**Issue by nbgl** [chengsoonong/mclass-sky#182]
This could be an alternative to KDE as a means of approximating uncertainties.

The derivation is
<img width="900" alt="screen shot 2017-06-22 at 22 45 49" src="https://user-images.githubusercontent.com/12507513/27434683-dffa0674-579c-11e7-8591-1dfd4e2aa7e8.png">
where Phi is the feature matrix.

**Comment by chengsoonong** [comment]
General knowledge for your education:
https://arxiv.org/pdf/1606.04820.pdf

**Comment by nbgl** [comment]
![uncertainty_test](https://user-images.githubusercontent.com/12507513/27673011-ce6cd39c-5ce1-11e7-8874-d806bb5f5bca.png)

The approximated uncertainties are very highly correlated to the actual uncertainties. The approximation tends to underestimate them by a (more-or-less) constant factor, but this is fine for GP as it (mostly) preserves order.

**Comment by chengsoonong** [comment]
How is GP uncertainty (the exact one) scaling with the size of the training set? Do we expect a larger training set to produce smaller uncertainty?

**Comment by nbgl** [comment]
I will make a separate issue of this.

I found a bug in my code that affected the above plot. The correct plot is
![uncertainty_test_fixed](https://user-images.githubusercontent.com/12507513/27689084-4d435b58-5d20-11e7-85ca-b2c086c96441.png)
The comments from above still hold.

Further, I have redrawn the learning curves (together with a changed kernel approximation).
![learning_curve](https://user-images.githubusercontent.com/12507513/27689103-69603748-5d20-11e7-92c2-b1424d29d606.png)
They look …weird. Why does the accuracy of GPs peak around 1000 and go down with more data? One possible explanation is that this is because I adjusted the hyperparameters by cross-validating on 1000 training points. It may be necessary to perform separate cross-validation for each number of training points. This sounds slow. Let’s discuss it tomorrow.
